Improve creation of origins and behaviors in cloudFront event

FanFataL opened the pull request:
## What did you implement

1. This Pull Request fix the issue with only one behavior with `PathPattern`. To solve this Im creating additional behavior without lambda functions and without `PathPattern` (It closes #7151)
2. For creating new behavior (default one) we need proper origin for this - if its only one across the stack we are using this, but if its more we need to point which origin is the default one. For this I've added new options to the cloudfront event `defaultOrigin: true`. You don't need to setup this when one of your behavior is without `PathPattern`
3. I've realized that there is no option right now to modify behavior config (eg If i need some custom headers to be forwarded). For this I've added new option to cloudfront event `behavior` - its a object defined by the user and its merged with our generated behavior. (it partialy resolve #6843)
4. I've spotted another bug with checking existing origins. If you want to create two origins to the same s3 server but one trought `s3://` protocol and second one trough `http://` protocol it was impossible. Also the naming based on the funciton name is wrong since one function can be attach to many origins.
5. I've added checking for gotchas to throw an error before we are trying to update the cloudformation stack

For every bug mentioned above Ive created test case for it to cover it.
A also update documentation which partialy resolve this issue #6785

## How can we verify it
<details>
<summary>1 - PathPattern</summary>
<!-- comment-->

```
functions:
  myLambdaAtEdge:
    handler: myLambdaAtEdge.handler
    events:
      - cloudFront:
          eventType: viewer-response
          pathPattern: /files*
          origin: s3://bucketname.s3.amazonaws.com/files
```

</details>
<details>
<summary>2 - DefaultOrigin</summary>
<!-- comment-->

```
functions:
  myLambdaAtEdge:
    handler: myLambdaAtEdge.handler
    events:
      - cloudFront:
          eventType: viewer-response
          pathPattern: /files*
          origin: s3://bucketname.s3.amazonaws.com/files
  myLambdaAtEdge2:
    handler: myLambdaAtEdge.handler
    events:
      - cloudFront:
          eventType: viewer-response
          pathPattern: /files2*
          defaultOrigin: true
          origin: s3://bucketname.s3.amazonaws.com/files2
```

</details>
<details>
<summary>3 - Extend behavior</summary>
<!-- comment-->

```
functions:
  myLambdaAtEdge:
    handler: myLambdaAtEdge.handler
    events:
      - cloudFront:
          eventType: viewer-response
          pathPattern: /files*
          behavior:
            AllowedMethods:
              - "GET"
              - "HEAD"
              - "OPTIONS"
              - "PUT"
              - "PATCH"
              - "POST"
              - "DELETE"
            CachedMethods:
              - "GET"
              - "HEAD"
              - "OPTIONS"
          origin: s3://bucketname.s3.amazonaws.com/files
```

</details>
<details>
<summary>4 - Origin bug</summary>
<!-- comment-->

```
functions:
  myLambdaAtEdge:
    handler: myLambdaAtEdge.handler
    events:
      - cloudFront:
          eventType: viewer-response
          origin: s3://bucketname.s3.amazonaws.com/files
  mySecondLambdaAtEdge:
    handler: myLambdaAtEdge.handler
    events:
      - cloudFront:
          eventType: origin-response
          pathPattern: /files*
          origin: http://bucketname.s3.amazonaws.com/files
```

</details>
<details>
<summary>5 - Throw an error on gotchas</summary>
<!-- comment-->

```
functions:
  myLambdaAtEdge:
    handler: myLambdaAtEdge.handler
    events:
      - cloudFront:
          eventType: viewer-request
          origin: s3://bucketname.s3.amazonaws.com/files
      - cloudFront:
          eventType: viewer-request
          origin: s3://bucketname.s3.amazonaws.com/other
```

</details>
<details>
<summary>Ultimate serverless configuration to check</summary>
<!-- comment-->

```
service: ultimate-cf-test

provider:
  name: aws
  runtime: nodejs10.x
  region: us-east-1
  versionFunctions: true

package:
  excludeDevDependencies: false
  exclude:
    - '**'
  include:
    - ./index.js

custom:
  cf:
    myWebsiteOrigin1:
      DomainName: serverless.com
      OriginPath: /framework
      CustomOriginConfig:
        OriginKeepaliveTimeout: 5
        OriginReadTimeout: 30
        OriginProtocolPolicy: match-viewer
    myWebsiteOrigin2: s3://bucketname.s3-eu-central-1.amazonaws.com/
    myWebsiteOrigin3: https://bucketname.s3-eu-central-1.amazonaws.com/
    myWebsiteOrigin4:
      DomainName:
        Fn::Join:
          - ''
          - - Ref: ApiGatewayRestApi
            - ".execute-api."
            - Ref: AWS::Region
            - ".amazonaws.com"
      OriginPath: /first
      CustomOriginConfig:
        OriginProtocolPolicy: match-viewer
    myWebsiteOrigin5:
      DomainName:
        Fn::Join:
          - ''
          - - Ref: ApiGatewayRestApi
            - ".execute-api."
            - Ref: AWS::Region
            - ".amazonaws.com"
      OriginPath: /second
      CustomOriginConfig:
        OriginProtocolPolicy: match-viewer
    myBehavior1:
      ForwardedValues:
        Cookies: {
          Forward: 'none'
        }
        QueryString: true
        Headers:
          - Authorization
          - Access-Control-Request-Headers
          - Access-Control-Request-Method
      Compress: true
      AllowedMethods:
        - "GET"
        - "HEAD"
        - "OPTIONS"
        - "PUT"
        - "PATCH"
        - "POST"
        - "DELETE"
      CachedMethods:
        - "GET"
        - "HEAD"
        - "OPTIONS"
    myBehavior2:
      Compress: true
      ViewerProtocolPolicy: redirect-to-https

functions:
  firstFunction:
    handler: index.handler
    events:
      - http: 'GET /first'
      - cloudFront:
          eventType: origin-request
          origin: ${self:custom.cf.myWebsiteOrigin1}
      - cloudFront:
          pathPattern: /first/*
          eventType: origin-request
          origin: ${self:custom.cf.myWebsiteOrigin2}
          behavior: ${self:custom.cf.myBehavior1}
      - cloudFront:
          pathPattern: /first2/*
          eventType: viewer-request
          origin: ${self:custom.cf.myWebsiteOrigin3}
          behavior: ${self:custom.cf.myBehavior2}
  secondFunction:
    handler: index.handler
    events:
      - http: 'GET /second'
      - cloudFront:
          eventType: origin-response
          origin: ${self:custom.cf.myWebsiteOrigin1}
      - cloudFront:
          pathPattern: /second/*
          eventType: origin-response
          origin: ${self:custom.cf.myWebsiteOrigin2}
          behavior: ${self:custom.cf.myBehavior1}
      - cloudFront:
          pathPattern: /second2/*
          eventType: viewer-response
          origin: ${self:custom.cf.myWebsiteOrigin3}
          behavior: ${self:custom.cf.myBehavior2}
  thirdFunction:
    handler: index.handler
    events:
      - cloudFront:
          eventType: viewer-request
          origin: ${self:custom.cf.myWebsiteOrigin1}
  fourthFunction:
    handler: index.handler
    events:
      - cloudFront:
          eventType: viewer-response
          origin: ${self:custom.cf.myWebsiteOrigin1}
      - cloudFront:
          pathPattern: /fourth/*
          eventType: viewer-response
          origin: ${self:custom.cf.myWebsiteOrigin4}
          behavior: ${self:custom.cf.myBehavior2}
  fifthFunction:
    handler: index.handler
    events:
      - cloudFront:
          pathPattern: /fifth/*
          eventType: viewer-response
          origin: ${self:custom.cf.myWebsiteOrigin5}
          behavior: ${self:custom.cf.myBehavior2}
```

</details>

## Todos

<details>
<summary>Useful Scripts</summary>
<!-- You might want to use the following scripts to streamline your development workflow -->

- `npm run test:ci` --> Run all validation checks on proposed changes
- `npm run lint:updated` --> Lint all the updated files
- `npm run lint:fix` --> Automatically fix lint problems (if possible)
- `npm run prettier-check:updated` --> Check if updated files adhere to Prettier config
- `npm run prettify:updated` --> Prettify all the updated files

</details>

- [x] Write and run all tests
- [x] Write documentation
- [x] Create ultimate serverless configuration to check
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

medikoo commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -545,8 +545,13 @@ module.exports = {
   },
-  getCloudFrontOriginId(functionName, path) {
-    let originId = this.getNormalizedFunctionName(functionName);
+  getCloudFrontOriginId(isS3Origin, domain, path) {
```
It'll probably be nicer to change signature simply to `originObj` and resolve needed values from it.

medikoo commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -545,8 +545,13 @@ module.exports = {
     return `${this.getLambdaLogicalId(functionName)}InvokePermission`;
   },
-  getCloudFrontOriginId(functionName, path) {
-    let originId = this.getNormalizedFunctionName(functionName);
```
Wouldn't it be breaking to change the originId, for existing stacks?

Maybe it'll be good to maintain same resolution for typical cases, and introduce new resolution for those which were not covered well (?)

FanFataL replied:
I've made a real tests from existing stack with old names to the new one and its only required CloudFront update - not breaking anything.

The goal of naming lib is ability to cross generate name for resources, if we add two different logic for generating names it wil be hard to achive that.

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -179,23 +184,35 @@ class AwsCompileCloudFrontEvents {
 
+            if (event.cloudFront.defaultOrigin === true) {
+              if (defaultOrigin !== undefined && defaultOrigin !== origin) {
+                throw new Error(
```
It should be `ServerlessError` (that way it'll be propagated as _user error_)

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -179,23 +184,35 @@ class AwsCompileCloudFrontEvents {
               },
+            };
+            if (_.isObject(event.cloudFront.behavior)) {
+              behavior = _.assignIn(behavior, event.cloudFront.behavior);
```
Let's use native `Object.assign`

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -179,23 +184,35 @@ class AwsCompileCloudFrontEvents {
+              behavior = _.assignIn(behavior, event.cloudFront.behavior);
+            }
+
+            behavior = _.assignIn(behavior, {
```
Same here

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -231,9 +248,9 @@ class AwsCompileCloudFrontEvents {
 
     // sort that first is without PathPattern if available
-    behaviors.sort((a, b) => {
+    behaviors = _.sortBy(behaviors, b => {
```
Why it was changed to `_.sortBy`? (currently our policy to resort to lodash, only if same operation cannot be achieved natively in clean way)

FanFataL replied:
There was a problem with this issue https://github.com/nodejs/node/issues/24294 on node12 because PathPatterns wasnt tested at all and bug wasnt spotted. Ive prepared fixed native sort function.

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -245,6 +262,41 @@ class AwsCompileCloudFrontEvents {
 
+      // Check if all behaviors got unique pathPatterns
+      if (behaviors.length !== _.uniqBy(behaviors, 'PathPattern').length) {
+        throw new Error('Found more than one behavior with the same PathPattern');
```
`ServerlessError`

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -245,6 +262,41 @@ class AwsCompileCloudFrontEvents {
+          );
+        })
+      ) {
+        throw new Error(
```
`ServerlessError`

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -245,6 +262,41 @@ class AwsCompileCloudFrontEvents {
+        let origin = defaultOrigin;
+        if (origin === undefined) {
+          if (origins.length > 1) {
+            throw new Error(
```
`ServerlessError`

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -245,6 +262,41 @@ class AwsCompileCloudFrontEvents {
 
+      // Check if all behaviors got unique pathPatterns
+      if (behaviors.length !== _.uniqBy(behaviors, 'PathPattern').length) {
+        throw new Error('Found more than one behavior with the same PathPattern');
```
Aren't those validation errors picked already by `CloudFormation.validateTemplate` SDK command, which we run internally?

FanFataL replied:
Unfortunatelly not

FanFataL commented:
> Great thanks @FanFataL it looks as great and notable improvement!
> 
> I've added some comments. Do you think it's possible to configure an effective integration tests for that (we host them here: https://github.com/serverless/serverless/tree/53a18cbff6d3d2d6698e98cf0dd8a7eba21fdf58/tests Or due to CloudFront nature it's hard to test it reasonably quickly that way?
> 
> It'll be great to have that covered by real tests that involve AWS stack.

Yes its possible, I will prepare some integrations tests but creation and removing stack will took about an hour - if its acceptable then no problem.

Additionally I've added checking proper runtime for Lambda@Edge and Ive created for this dedicated method in provider.
You can review.

medikoo commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -545,8 +545,17 @@ module.exports = {
-  getCloudFrontOriginId(functionName, path) {
-    let originId = this.getNormalizedFunctionName(functionName);
+  getCloudFrontOriginId(originObj) {
+    const isS3Origin = originObj.S3OriginConfig !== undefined;
```
Let's do `Boolean(originObj.S3OriginConfig)` (we also should discard values as `null` or `false`)

medikoo commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -545,8 +545,17 @@ module.exports = {
+    let originId = isS3Origin ? 's3' : 'custom';
+    const domainName = _.isString(domain)
+      ? domain
+      : this.normalizeNameToAlphaNumericOnly(JSON.stringify(domain));
```
What is the use case for configuring `originObj.DomainName` as an object?

FanFataL replied:
When you are referencing to the cloudformation variables/functions eg:
```
    origin:
      DomainName:
        Fn::Join:
          - ''
          - - Ref: ApiGatewayRestApi
            - ".execute-api."
            - Ref: AWS::Region
            - ".amazonaws.com"

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -7,6 +7,7 @@ const { ServerlessError } = require('../../../../../../classes/Error');
 
 const originLimits = { maxTimeout: 30, maxMemorySize: 3008 };
 const viewerLimits = { maxTimeout: 5, maxMemorySize: 128 };
+const supportedRuntimes = ['nodejs6.10', 'nodejs8.10', 'nodejs10.x', 'python3.7'];
```
Is `nodejs12.x` not yet supported?

Hardcoding runtimes implies a maintenance issue (we need to introduce code changes on every change in that).
Can we leave it for AWS validation, or maybe there's some way to confirm on that by issuing an AWS request?

FanFataL replied:
Node12 is supported in normal lambdas but tts not supported as a Lambda@Edge

FanFataL replied:
Ok - I removed this validation and move this information to the documentation.

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -177,25 +188,36 @@ class AwsCompileCloudFrontEvents {
-              o => o.DomainName === origin.DomainName && o.OriginPath === origin.OriginPath
-            );
+            let origin = createOrigin(event.cloudFront.origin, this.provider.naming);
+            const existingOrigin = _.find(origins, o => o.Id === origin.Id);
```
Let's use native `.find`

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -177,25 +188,36 @@ class AwsCompileCloudFrontEvents {
             }
 
+            if (event.cloudFront.defaultOrigin === true) {
+              if (defaultOrigin !== undefined && defaultOrigin !== origin) {
```
Let's make it `defaultOrigin && defaultOrigin !== origin` (otherwise values as `null` are treated as if `defaultOrigin` is set)

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -245,6 +270,41 @@ class AwsCompileCloudFrontEvents {
+      // DefaultCacheBehavior does not support PathPattern property
+      if (behaviors[0].PathPattern) {
+        let origin = defaultOrigin;
+        if (origin === undefined) {
```
`if (!origin)`

medikoo commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -454,6 +454,20 @@ class AwsProvider {
 
+  getRuntimeSourceValue() {
+    const values = this.getValues(this, [
+      ['options', 'runtime'],
```
`runtime` is not a thing that should be overriden via CLI option.

medikoo commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -454,6 +454,20 @@ class AwsProvider {
+  getRuntimeSourceValue() {
+    const values = this.getValues(this, [
+      ['options', 'runtime'],
+      ['serverless', 'config', 'runtime'],
```
Same, we do not support it as top level config option, and I don't think it should change with this PR

FanFataL replied:
ok

medikoo commented on lib/plugins/package/lib/packageService.js:
```diff
@@ -23,6 +23,17 @@ module.exports = {
+    return (
+      runtime ||
+      this.options.runtime ||
+      this.serverless.config.runtime ||
```
Same concerns apply here.

(Otherwise it's a good call, to unify runtime resolution)

medikoo commented:
> Yes its possible, I will prepare some integrations tests but creation and removing stack will took about an hour - if its acceptable then no problem.

In that case, we may leave it for now. An hour long integration tests cannot be easily automated

FanFataL commented:
> > Yes its possible, I will prepare some integrations tests but creation and removing stack will took about an hour - if its acceptable then no problem.
> 
> In that case, we may leave it for now. An hour long integration tests cannot be easily automated

You can consider to do integration tests on LocalStack Pro - it should be much faster to have it locally.

medikoo commented:
> You can consider to do integration tests on LocalStack Pro - it should be much faster to have it locally.

That's definitely a great idea, we'll check that out

medikoo commented on docs/providers/aws/events/cloudfront.md:
```diff
@@ -66,11 +68,39 @@ module.exports.handler = (event, context, callback) => {
+      - cloudFront:
+          eventType: viewer-response
+          pathPattern: /files*
+          defaultOrigin: true
```
Let's make it `isDefaultOrigin` to signal it's a boolean and not string

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -177,25 +178,36 @@ class AwsCompileCloudFrontEvents {
               origin = extendDeep(existingOrigin, origin);
             }
 
+            if (event.cloudFront.defaultOrigin === true) {
```
`if (event.cloudFront.defaultOrigin)`

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -177,25 +178,36 @@ class AwsCompileCloudFrontEvents {
                 QueryString: false,
               },
+            };
+            if (_.isObject(event.cloudFront.behavior)) {
```
If I understand correct, object representation it's only one we accept (there's no string version).

In that case, it's better written as `if (event.cloudFront.behavior)`

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -97,6 +97,7 @@ class AwsCompileCloudFrontEvents {
 
     const origins = [];
     const behaviors = [];
+    let isDefaultOrigin;
```
This one should be kept as `defaultOrigin`, as it hosts origin string and not a boolean

jplock commented:
This change breaks the ability to do this in a CloudFormation template:

```
        Origins:
          - Id: web
            DomainName:
              'Fn::GetAtt':
                - WebBucket
                - DomainName
            S3OriginConfig:
              OriginAccessIdentity:
                'Fn::Join':
                  - '/'
                  - - 'origin-access-identity'
                    - cloudfront
                    - Ref: CloudFrontWebAccess
```

Serverless fails with:

```
Type Error ---------------------------------------------
 
  TypeError: Cannot read property 'replace' of undefined
      at Object.normalizeNameToAlphaNumericOnly (/Users/me/app/cdn/node_modules/serverless/lib/plugins/aws/lib/naming.js:32:36)
      at Object.getCloudFrontOriginId (/Users/me/app/cdn/node_modules/serverless/lib/plugins/aws/lib/naming.js:556:14)
      at createOrigin (/Users/me/app/cdn/node_modules/serverless/lib/plugins/aws/package/compile/events/cloudFront/index.js:144:20)
...
```

FanFataL commented:
> This change breaks the ability to do this in a CloudFormation template:
> 
> ```
>         Origins:
>           - Id: web
>             DomainName:
>               'Fn::GetAtt':
>                 - WebBucket
>                 - DomainName
>             S3OriginConfig:
>               OriginAccessIdentity:
>                 'Fn::Join':
>                   - '/'
>                   - - 'origin-access-identity'
>                     - cloudfront
>                     - Ref: CloudFrontWebAccess
> ```
> 
> Serverless fails with:
> 
> ```
> Type Error ---------------------------------------------
>  
>   TypeError: Cannot read property 'replace' of undefined
>       at Object.normalizeNameToAlphaNumericOnly (/Users/me/app/cdn/node_modules/serverless/lib/plugins/aws/lib/naming.js:32:36)
>       at Object.getCloudFrontOriginId (/Users/me/app/cdn/node_modules/serverless/lib/plugins/aws/lib/naming.js:556:14)
>       at createOrigin (/Users/me/app/cdn/node_modules/serverless/lib/plugins/aws/package/compile/events/cloudFront/index.js:144:20)
> ...
> ```
> 
> Because
> 
> ```
> getCloudFrontOriginId(originObj) {
>     const isS3Origin = Boolean(originObj.S3OriginConfig);
>     const domain = originObj.DomainName;
>     const path = originObj.OriginPath;
> 
>     let originId = isS3Origin ? 's3' : 'custom';
>     const domainName = _.isString(domain)
>       ? domain
>       : this.normalizeNameToAlphaNumericOnly(JSON.stringify(domain));
> 
>     originId = `${originId}/${domainName}`;
>     if (path) {
>       originId = `${originId}${path}`;
>     }
>     return originId;
>   },
> ```
> 
> the `domainName` isn't properly handling CloudFormation intrinsic functions

Hi, Thanks for comment.

The getCloudFrontOriginId has been called only with the function `cloudFront` event. Are you using some plugin to generate origins from the cloudformation template? 

It seems like `JSON.stringify(domain)` returns `undefined`

jplock commented:
I was defining my entire CloudFront distribution in a CloudFormation template and only used Serverless to hookup the Lambda@Edge functions.

FanFataL commented:
> I was defining my entire CloudFront distribution in a CloudFormation template and only used Serverless to hookup the Lambda@Edge functions.

Can you give example of function definition with the events how you definining it?

jplock commented:
@FanFataL , my serverless.yml looks like:

```
functions:
  defaultIndex:
    name: '${self:service}-${self:provider.stage}-origin-request-defaultIndex'
    handler: functions/default_index.handler
    memorySize: 128 # megabytes
    timeout: 1 # seconds
    events:
      - cloudFront:
          eventType: origin-request

  webSecurity:
    name: '${self:service}-${self:provider.stage}-origin-response-webSecurity'
    handler: functions/web_security.handler
    memorySize: 128 # megabytes
    timeout: 1 # seconds
    events:
      - cloudFront:
          eventType: origin-response
```

And then in a separate resources file, I'm defining the CloudFrontDistribution:

```
---
Resources:
  CloudFrontDistribution:
    Type: 'AWS::CloudFront::Distribution'
    Properties:
      DistributionConfig:
...
```

FanFataL commented:
Heh, I've checked the code before my change and it was working your way accidently :)

The only thing you can do right now is moving origin creation config into `cloudFront` event configuration in serverless.yml like described here: https://serverless.com/framework/docs/providers/aws/events/cloudfront/#simple-event-definition

You can also create feature request to support origins created in the cf template.
